packages/eslint-plugin/src/rules/member-ordering.ts

```diff
@@ -307,10 +311,16 @@ function getMemberRawName(
   return name;
 }
 
+function getEnumMemberName(member: TSESTree.TSEnumMember): string | null {
+  const enumID = member.id as unknown as Record<string, string | undefined>;
```

Comment:
The fact that this assertion was needed is an indication that the type system isn't being respected in this code. I see from going to definition on `id` that it can be type `PropertyNameComputed` or `PropertyNameNonComputed`. TIL this is possible!

```ts
const myConstant = 'd';

enum Computed {
   ['a'] = 0,
   ['b' + 'c'] = 1, // Type Error: Computed property names are not allowed in enums.
   [myConstant] = 2, // Type Error: Computed property names are not allowed in enums.
}
```

Instead of using an assertion, this code should check the type of `member.id` and respond accordingly.

Note that even though TypeScript will emit a type error for the two computed names, users may still run this rule on the code via IDE plugins. The rule will have to handle the case of a computed property that it can't know the value of, even if it returns a falsy value as it does now.

Reply:
I'm struggling to understand. Do you want me to change it to be this?

```ts
function isIdentifier(): is ts.Identifier {
   ???
}

function getEnumMemberName(
  member:
    | TSESTree.TSEnumMemberComputedName
    | TSESTree.TSEnumMemberNonComputedName,
): string | null {
  if (isIdentifier(member.id)) {
    return null;
  }
  
  return member.id.name === "" ? null : member.id.name;
}
```

Reply:
Oh, I think you mean this, right?

```ts
function getEnumMemberName(
  member:
    | TSESTree.TSEnumMemberComputedName
    | TSESTree.TSEnumMemberNonComputedName,
): string | null {
  if (!("name" in member.id)) {
    return null;
  }

  return member.id.name === "" ? null : member.id.name;
}
```

Reply:
Sorry I wasn't very clear 😞! We typically check whether the node's `type` property is equal to some member of `AST_NODE_TYPES`. If the `type` can be multiple things there's a good chance some existing `isSomeNodeType` type predicate function exists in `node-utils.ts`. That file has a lot of functions you can read as reference to see what kinds of checks we typically use on `TSESTree.*` and `ts.*` nodes types.

For this one, try copying in regular enums and the code snippet I pasted to the typescript-eslint.io/play playground and setting the AST Viewer to ESTree [[link for convenience](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&showAST=es&code=MYewdgzgLgBAtgTwMLmgQzLAvDA5AE1wG4BYAKHIFMwBXOGFOABxqkvxgG9yZeYAPGDgC0ARgA0PXgG1caXAF0hMAAySyfWQCNcMANR5gi5RJgB6MzAAqCJpRgBRAE5OQTgFwMQzVuxhNXOycoBBgwNDhKCBg0J3swEFg0ABtkkAB3PwBLMBhqOggAOikYaUQUSCgMKCUcACZxc0sbO0cXN09GFjYOAJAgkLCIqJi4sMSY1Izs3Py4IvIAX3IgA&rules=N4XyA&tsConfig=N4XyA)]. It'll show the following types for these `member.id`s:

* `a`: `Identifier`, and `member.computed` is `false`
* `['a']`: `Literal`, and `member.computed` is `true`
* `['b' + 'c']`: `BinaryExpression`, and `member.computed` is `true`
* `[myConstant]`: `Identifier`, and `member.computed` is `true`

Using that info, I think `getEnumMemberName` should probably check for a few cases:
* If `member.computed` is `false`: only return a name if `member.id.type` is `AST_NODE_TYPES.Identifier`
* * If `member.computed` is `true`: only return a name if `member.id.type` is `AST_NODE_TYPES.Literal`

Does that make sense & seem reasonable to you?

By the way, we normally use `undefined` to indicate blank values, not `null`. `undefined` is the default _"this doesn't exist"_ in JS and it's confusing having two.